### PR TITLE
[GEOT-6871] AppSchema defining nested attribute with root attribute filter might return 0 on count

### PR DESCRIPTION
**Pull request by taba90**
Jira ticket https://osgeo-org.atlassian.net/browse/GEOT-6871
Integration tests provided here https://github.com/geoserver/geoserver/pull/4973
# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [x] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [x] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [x] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [x] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java [hunk 1]
```diff
@@ -470,10 +465,7 @@ public String selectSQL(
                 }
 
                 joinClause.append(" = ");
-                String fromTypeName;
-                if (isCount && join.getJoinedTypeName() != null)
-                    fromTypeName = join.getJoinedTypeName();
-                else fromTypeName = curTypeName;
+                String fromTypeName = curTypeName;
```
Comment (fernandor777) [inline modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java:468]: Question: why we remove the joined type name use case?

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/AppSchemaDataAccess.java [hunk 2]
```diff
@@ -347,25 +341,22 @@ private boolean canCount(
...
             throws IOException {
-        List<FeatureChainedAttributeDescriptor> featureChainDescriptors =
-                getFeatureChainedDescriptorsFromFilter(mapping, propertyNames);
         // wrap the query in the Joining one to have the unmapping visitor
         // setting the ids
         JoiningQuery unrollQuery =
                 (JoiningQuery) unrollQuery(new JoiningQuery(targetQuery), mapping);
-        JoiningQuery joining = buildJoiningQuery(featureChainDescriptors, unrollQuery, mapping);
-        return source.getCount(joining);
+        unrollQuery.setRootMapping(mapping);
```
Comment (fernandor777) [inline modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/AppSchemaDataAccess.java:358]: Joining query instance is not used anymore, is this safe for legacy functionality?  Please test these changes with postgis integration tests.

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java [hunk 3]
```diff
@@ -1458,22 +1448,33 @@ protected int getCountInternal(Query query) throws IOException {
...
+        MultipleValueExtractor extractor = new MultipleValueExtractor();
         jQuery.getRootMapping().getFeatureIdExpression().accept(extractor, null);
         String[] attributes = extractor.getAttributeNames();
         // the id from the app-schema configuration
         String idMapping = attributes.length > 0 ? extractor.getAttributeNames()[0] : null;
         boolean idsColumnEquals = idColumnName.equals(idMapping);
-        boolean isJoining = jQuery.getQueryJoins().size() > 0;
+        jQuery.getFilter().accept(extractor, null);
+        boolean isNestedFilter =
+                NestedFilterToSQL.isNestedFilter(jQuery.getFilter())
```
Comment (fernandor777) [inline modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java:1465]: This boolean building block is large enough to be a private method explaining how this work.

**Review by fernandor777 — CHANGES_REQUESTED** [review]
We need to double check if we are introducing breaking changes for joining queries on App-Schema.  Please read my questions/comments.

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java (hunk 1 above)
Comment (taba90) [reply modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java:469]: introduced by me in a previous commit related to the count functionalities. It was not actually needed so I'm removed

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/AppSchemaDataAccess.java (hunk 2 above)
Comment (taba90) [reply modules/extension/app-schema/app-schema/src/main/java/org/geotools/data/complex/AppSchemaDataAccess.java:358]: Same as above.

File: modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java (hunk 3 above)
Comment (taba90) [reply modules/extension/app-schema/app-schema/src/main/java/org/geotools/appschema/jdbc/JoiningJDBCFeatureSource.java]: fixed

**Comment by taba90** [comment]
thanks for your review @fernandor777  I replied to the comments and applied the feedback

**Comment by taba90** [comment]
unrelated failure on Downstream integration build Error:  `Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (default-test) on project gwc-core: There are test failures.`
unrelated failure on linux builds: `Error:  Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.22.1:test (default-test) on project gt-jdbc-h2: There are test failures.`

**Pull request merged**